File: Model/satellite.cpp
#include "satellite.h"

Satellite::Satellite(double r, double ts, double dm, unsigned int e, double vRot, double semiA, const Pianeta &p):
    Orbitante(r,ts,dm,e,vRot, semiA), pianeta(p){}

Satellite::Satellite(const Orbitante& o, const Pianeta& p): Orbitante(o), pianeta(p) {}

const Pianeta &Satellite::getPianeta() const{
    return pianeta;
}

int Satellite::rivoluzioniAnnue() const{
    return (pianeta.periodoOrbitale().AnniFraz()*8760+pianeta.periodoOrbitale().hour())/(periodoOrbitale().AnniFraz()*8760 + periodoOrbitale().hour());
}

DataOraTerrestre Satellite::periodoOrbitale() const{
    long double num = 4*pow(pi,2)*pow(getAsse()/6.684587123e-12,3);
    long double den = G*(Massa()*pow(10,12)+pianeta.Massa()*pow(10,12));
    long double result = sqrt(num/den)*(3.168753579e-08)*365;
    return DataOraTerrestre(result*86400);
}


bool Satellite::rotazioneSincrona() const{ //dato che l'approssimazione puo non essere perfetta, confronto solo le ore
    return Giorno().hour() + Giorno().Giorni()*24 <= (periodoOrbitale().hour() + periodoOrbitale().Giorni()*24 + 0.1*(periodoOrbitale().Giorni()*24 + periodoOrbitale().hour())) &&
            Giorno().hour() + Giorno().Giorni()*24 >= (periodoOrbitale().hour() + periodoOrbitale().Giorni()*24 - 0.1*(periodoOrbitale().Giorni()*24 + periodoOrbitale().hour())) ;
}

double Satellite::velOrbitale() const{
    return sqrt((G*pianeta.Massa())/(pianeta.getRaggio()*pow(10,3)+getAsse()*1.5e+11))*pow(10,3);
}

double Satellite::distanzaSole() const{
    return pianeta.distanzaSole()+getAsse();
}


Satellite Satellite::operator+(const Satellite& o) const{
    Satellite aux(*this);
    aux.fusione(o);
    return aux;
}

File: Java/src/astrokalk/Satellite.java
package astrokalk;

import java.time.Duration;

import static java.lang.Math.pow;
import static java.lang.Math.sqrt;

public class Satellite extends Orbitante {

    private Pianeta pianeta;


    public Satellite(double r, double ts, double dm, int e, double vRot, double semiA, Pianeta p) throws EccInput {
        super(r, ts, dm, e, vRot, semiA);
        pianeta = p;
    }

    public Satellite(Orbitante o, Pianeta p) {
        super(o);
        pianeta = p;
    }

    public Satellite(Satellite s) {
        super(s);
        pianeta = s.pianeta;
    }

    public Pianeta getPianeta() {
        return pianeta;
    }

    public double rivoluzioniAnnue() {
        return ((double) pianeta.periodoOrbitale().toHours()) / (double) periodoOrbitale().toHours();
    }

    public Duration periodoOrbitale() {
        double num = 4 * pow(pi, 2) * pow(getAsse() / 6.684587123e-12, 3);
        double den = G * (massa() * pow(10, 12) + pianeta.massa() * pow(10, 12));
        double result = sqrt(num / den) * (3.168753579e-08) * 365;
        return Duration.ofSeconds((long) (result * 86400));
    }


    public boolean rotazioneSincrona() { //dato che l'approssimazione puo non essere perfetta, confronto solo le ore
        return Giorno().toHours() <= (periodoOrbitale().toHours() + 0.1 * periodoOrbitale().toHours()) &&
                Giorno().toHours() >= (periodoOrbitale().toHours() - 0.1 * periodoOrbitale().toHours());
    }

    public double velOrbitale() {
        return sqrt((G * pianeta.massa()) / (pianeta.getRaggio() * pow(10, 3) + getAsse() * 1.5e+11)) * pow(10, 3);
    }

    public double distanzaSole() {
        return pianeta.distanzaSole() + getAsse();
    }


    public Satellite somma(Satellite o) {
        Satellite aux = new Satellite(this);
        aux.fusione(o);
        aux.pianeta = o.pianeta;
        return aux;
    }

}
